File: java/3093. Longest Common Suffix Queries.java
class Solution {
    public int[] stringIndices(String[] wordsContainer, String[] wordsQuery) {
        Trie db = new Trie();
        for (int i = 0; i < wordsContainer.length; i++) {
            db.add(wordsContainer[i], i);
        }
        int[] answer = new int[wordsQuery.length];
        for (int i = 0; i < wordsQuery.length; i++) {
            String query = wordsQuery[i];
            answer[i] = db.find(query);
        }
        return answer;
    }
}

class Trie {
    Trie[] children;
    int index;
    int length;
    Trie() {
        children = new Trie[26];
        index = Integer.MAX_VALUE;
        length = Integer.MAX_VALUE;
    }
    void add(String word, int index) {
        Trie current = this;
        if (current.length > word.length()) {
            current.length = word.length();
            current.index = index;
        }
        for (int i = word.length()-1; i >= 0; i--) {
            char c = word.charAt(i);
            if (current.children[c-'a'] == null) {
                current.children[c-'a'] = new Trie();
            }
            current = current.children[c-'a'];
            if (current.length > word.length()) {
                current.length = word.length();
                current.index = index;
            }
        }
    }
    int find(String q) {
        Trie current = this;
        for (int i = q.length()-1; i >= 0; i--) {
            char c = q.charAt(i);
            if (current.children[c-'a'] == null) break;
            current = current.children[c-'a'];
        }
        return current.index;
    }
}

File: python/3093. Longest Common Suffix Queries.py
class Solution:
    def stringIndices(self, wordsContainer: List[str], wordsQuery: List[str]) -> List[int]:
        db = Trie()
        for i, word in enumerate(wordsContainer):
            db.add(word, i)
        return [db.find(word) for word in wordsQuery]

class Trie:
    def __init__(self):
        self.children = {}
        self.index = +inf
        self.length = +inf
        
    def add(self, word, index):
        current = self
        if current.length > len(word):
            current.length = len(word)
            current.index = index
        for i in range(len(word)):
            c = word[len(word) - i - 1]
            if c not in current.children:
                current.children[c] = Trie()
            current = current.children[c]
            if current.length > len(word):
                current.length = len(word)
                current.index = index
        
    def find(self, word):
        current = self
        for i in range(len(word)):
            c = word[len(word) - i - 1]
            if c not in current.children:
                break
            current = current.children[c]
        return current.index
